### PR TITLE
VERTEX_DATA: add format, offset, stride parameters

### PR DESCRIPTION
**Pull request by asuonpaa**

### CODE REVIEW

File: src/amberscript/parser.cc [hunk 1]
```diff
@@ -1301,20 +1301,55 @@ Result Parser::ParsePipelineVertexData(Pipeline* pipeline) {
   const uint32_t location = token->AsUint32();
 
   InputRate rate = InputRate::kVertex;
+  uint32_t offset = 0;
+  Format* format = buffer->GetFormat();
+  uint32_t stride = format->SizeInBytes();
```
Comment (paulthomson) [inline src/amberscript/parser.cc:1306]: This is kind of subtle but I think if somebody overrides the format, but not the stride, then the stride should be derived from the overridden format, not the format of the original buffer. I guess you can check if the stride was explicitly set after the loop (I.e. It is non-zero? Although I guess we should not allow an explicit 0 stride?) and, if not, set it based on the format. And I guess add a test for this as well.

**Review by paulthomson — COMMENTED** [review]
LGTM! I left a few comments. And lets get @dj2 to quickly review the feature addition by reviewing `amber_script.md`.

File: docs/amber_script.md [hunk 2]
```diff
@@ -588,8 +588,12 @@ when using dynamic buffers with BUFFER_ARRAY.
...
 ```groovy
   # Set |buffer_name| as the vertex data at location |val|. `RATE` defines the
-  # input rate for vertex attribute reading.
-  VERTEX_DATA {buffer_name} LOCATION _val_ [ RATE { vertex | instance } (default vertex) ]
+  # input rate for vertex attribute reading. `FORMAT` can be used to override
+  # the vertex buffer format specified by the |buffer_name|. `OFFSET` defines a
+  # byte offset used for the vertex data binding, and `STRIDE` overrides the
+  # default stride specified by the |buffer_name|.
+  VERTEX_DATA {buffer_name} LOCATION _val_ [ RATE { vertex | instance } (default vertex) ] \
+        [ FORMAT {format} ] [ OFFSET {offset} ] [ STRIDE {stride} ]
```
[the remark this answers is not in the document]
Comment (asuonpaa) [reply docs/amber_script.md]: done

File: src/amberscript/parser.cc (hunk 1 above)
Comment (asuonpaa) [reply src/amberscript/parser.cc]: done

**Review by dj2 — APPROVED** [review]

**Review by paulthomson — APPROVED** [review]
LGTM!

**Pull request merged**